middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -769,6 +768,37 @@ let close_named acc env ~let_bound_var (named : IR.named)
       ~current_region:(fst (Env.find_var env region))
       k
 
+type simplified_block_load =
+  | Unknown
+  | Not_a_block
+  | Block_but_cannot_simplify of Code_or_metadata.t Value_approximation.t
+  | Field_contents of Symbol.t
+
+let simplify_block_load acc body_env ~block ~field : simplified_block_load =
+  match find_value_approximation_through_symbol acc body_env block with
+  | Value_unknown -> Unknown
+  | Closure_approximation _ | Value_symbol _ | Value_int _ -> Not_a_block
+  | Block_approximation (approx, _alloc_mode) -> (
+    let approx =
+      Simple.pattern_match field
+        ~const:(fun const ->
+          match Reg_width_const.descr const with
+          | Tagged_immediate i ->
+            let i = Targetint_31_63.to_int i in
+            if i >= Array.length approx
+            then
+              Misc.fatal_errorf
```

Comment:
Isn't it possible for this to appear with GADT code.

Reply:
I haven't actually checked (I haven't got this branch checked out currently), but I would expect that with minimal changes you could get this to trigger.